File: examples/python/version_conflict.py
"""Protect a document with optimistic version checks."""

from __future__ import annotations

import tempfile

from dirdb import DirDB


def main() -> None:
    db = DirDB(tempfile.mkdtemp(prefix="dirdb-version-"))
    first = db.set("services/payment/config", {"enabled": False})
    second = db.set(
        "services/payment/config",
        {"enabled": True},
        expected_version=first,
    )
    print("updated to version:", second)

    try:
        db.set("services/payment/config", {"enabled": False}, expected_version=first)
    except RuntimeError as error:
        print("stale update rejected:", error)


if __name__ == "__main__":
    main()

File: examples/rust/basic/src/bin/version_conflict.rs
use dirdb_core::{DirDb, Error};
use serde_json::json;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let db = DirDb::open("example-state/rust/version-conflict")?;
    let first = db.set("services/payment/config", &json!({"enabled": false}), None)?;
    let second = db.set(
        "services/payment/config",
        &json!({"enabled": true}),
        Some(first.version),
    )?;
    println!("updated to version: {}", second.version);

    match db.set(
        "services/payment/config",
        &json!({"enabled": false}),
        Some(first.version),
    ) {
        Err(Error::VersionConflict { expected, actual, .. }) => {
            println!("stale update rejected: expected {expected}, actual {actual}");
        }
        Ok(_) => unreachable!("a stale update must not succeed"),
        Err(error) => return Err(error.into()),
    }

    Ok(())
}
